File: py/0819_Most_Common_Word.py
import sys
from typing import *
from collections import OrderedDict
from collections import deque
import re

class Solution:
    def mostCommonWord(self, paragraph: str, banned: List[str]) -> str:
        words = re.split(r"[!?',;. ]", paragraph)
        record = {"": -1}
        ret = ""
        max_times = 0

        for ban in banned:
            record[ban] = -1

        for w in words:
            w = w.lower()
            if w not in record:
                record[w] = 1
            else:
                if record[w] != -1:
                    record[w] += 1

            if record[w] > max_times:
                ret = w
                max_times = record[w]

        return ret

if __name__ == "__main__":
    solution = Solution()

    # Input: paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.", banned = ["hit"]
    # Output: "ball"
    paragraph = "Bob hit a ball, the hit BALL flew far after it was hit."
    banned = ["hit"]
    print(solution.mostCommonWord(paragraph, banned))

    # Input: paragraph = "a.", banned = []
    # Output: "a"
    paragraph = "a."
    banned = [""]
    print(solution.mostCommonWord(paragraph, banned))

    sys.exit(0)

File: cpp/0819_Most_Common_Word.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cmath>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Solution {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        map<string, int> record;
        int max_times = 0;
        string ret;

        for (auto b : banned) {
            record[b] = -1;
        }

        paragraph += ' ';
        string tmp = "";
        for (auto c : paragraph) {
            if (c == ' ' || c == '!' || c == '?' || c == '\'' || c == ',' || c == ';' || c == '.') {
                if (tmp != "") {
                    if (record.count(tmp) == 0) {
                        record[tmp] = 1;
                    } else {
                        if (record[tmp] != -1) {
                            record[tmp]++;
                        }
                    }

                    if (record[tmp] > max_times) {
                        max_times = record[tmp];
                        ret = tmp;
                    }

                    tmp = "";
                }
                continue;
            }

            tmp += tolower(c);
        }

        return ret;
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    string paragraph;
    vector<string> banned;

    // Input: paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.", banned = ["hit"]
    // Output: "ball"
    paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.";
    banned = {"hit"};
    cout << solution.mostCommonWord(paragraph, banned) << endl;

    // Input: paragraph = "a.", banned = []
    // Output: "a"
    paragraph = "a.";
    banned = {};
    cout << solution.mostCommonWord(paragraph, banned) << endl;

    return 0;
}
